File: leetcode/js/longest-harmonious-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findLHS = function (nums) {
  let result = 0;

  for (const num of nums) {
    if (nums.includes(num + 1)) {
      const count = nums.filter((n) => n === num || n === num + 1).length;
      result = Math.max(count, result);
    }
  }

  return result;
};

var findLHS = function (nums) {
  const numsMap = new Map();
  let result = 0;

  for (const num of nums) {
    numsMap.set(num, numsMap.has(num) ? numsMap.get(num) + 1 : 1);
  }

  for (const [key, val] of numsMap) {
    if (numsMap.has(key + 1)) {
      result = Math.max(numsMap.get(key) + numsMap.get(key + 1), result);
    }
  }

  return result;
};

File: leetcode/rust/longest-harmonious-subsequence.rs
impl Solution {
    pub fn find_lhs(nums: Vec<i32>) -> i32 {
        use std::collections::HashMap;

        let mut map = HashMap::new();
        let mut result = 0;

        for num in nums.into_iter() {
            match map.get(&num) {
                Some(&count) => map.insert(num, count + 1),
                None => map.insert(num, 1),
            };
        }

        for (key, val) in map.iter() {
            if map.contains_key(&(key + 1)) {
                result = result.max(val + map.get(&(key + 1)).unwrap());
            }
        }

        result
    }
}
